### PR TITLE
feat: YataiClient context to communicate with multiple Yatai instances

### PR DESCRIPTION
**Pull request by ssheng**
## What does this PR address?

The `.yatai.yaml` configuration under `BENTOML_HOME` already supports contexts to communicate with multiple Yatai instances through different API tokens today. However, the same functionality was not exposed to the CLI commands and YataiClient API. This PR exposes setting and selecting contexts in the CLI.

Yatai Log-in
```
> bentoml yatai login --api-token xxx --endpoint https://yatai1 --context yatai1
Successfully logged in as user "ssheng" in organization "yatai1".
> bentoml yatai login --api-token xxx --endpoint https://yatai2 --context yatai2
Successfully logged in as user "ssheng" in organization "yatai2".
```

Bento push & pull
```
> bentoml push iris_classifier:x67srpdqs25aicvj --context yatai1
> bentoml push iris_classifier:x67srpdqs25aicvj --context yatai2
```

Model push & pull
```
> bentoml models push iris_clf:7r5r5ftamc3dccvj --context=yatai1
> bentoml models push iris_clf:7r5r5ftamc3dccvj --context=yatai2
```

YataiClient API
```
yatai_client.push_bento(bento, context="yatai1")
yatai_client.pull_bento(bento_tag, context="yatai1")
yatai_client.push_model(model, context="yatai1")
yatai_client.pull_model(model_tag, context="yatai1")
```

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/bentoml/_internal/yatai_rest_api_client/config.py [hunk 1]
```diff
@@ -108,3 +112,16 @@ def get_current_context() -> YataiClientContext:
 def get_current_yatai_rest_api_client() -> YataiRESTApiClient:
     ctx = get_current_context()
     return ctx.get_yatai_rest_api_client()
```
Comment (ShaohongBai) [inline src/bentoml/_internal/yatai_rest_api_client/config.py:114]: @ssheng thank you again for speedy work on this! It is indeed what we expect. 

Just a slight minder, depends how restrict your code space is, the highlighted function will be redundant after your new implementation, I think.

Comment (ssheng) [reply src/bentoml/_internal/yatai_rest_api_client/config.py]: Thanks for the suggestion, @ShaohongBai.

**Comment by ssheng** [comment]
> I'm thinking maybe for `push` and `pull` we can be more explicit
> 
> ```shell
> bentoml models push ... --yatai-context
> ```
> 
> whereas for `bentoml yatai` `--context` works?

Make sense but in the case of BentoCloud, I worry that the `yatai` prefix maybe irrelevant or misleading.

**Comment by aarnphm** [comment]
> Make sense but in the case of BentoCloud, I worry that the `yatai` prefix maybe irrelevant or misleading.

How about ``bentoml models push --remote-context``?

File: src/bentoml_cli/yatai.py [hunk 2]
```diff
@@ -20,7 +20,13 @@ def yatai_cli():
         "--endpoint", type=click.STRING, help="Yatai endpoint, i.e: https://yatai.com"
     )
     @click.option("--api-token", type=click.STRING, help="Yatai user API token")
-    def login(endpoint: str, api_token: str) -> None:  # type: ignore (not accessed)
+    @click.option(
+        "--context",
```
Comment (aarnphm) [inline src/bentoml_cli/yatai.py:24]:
```suggestion
        "--client-context",
```

File: src/bentoml_cli/yatai.py [hunk 3]
```diff
@@ -20,7 +20,13 @@ def yatai_cli():
...
+        type=click.STRING,
+        help="Yatai context name for the endpoint and API token",
+        default=default_context_name,
+    )
+    def login(endpoint: str, api_token: str, context: str) -> None:  # type: ignore (not accessed)
```
Comment (aarnphm) [inline src/bentoml_cli/yatai.py:29]:
```suggestion
    def login(endpoint: str, api_token: str, client_context: str) -> None:  # type: ignore (not accessed)
```

File: src/bentoml_cli/yatai.py [hunk 4]
```diff
@@ -40,7 +46,7 @@ def login(endpoint: str, api_token: str) -> None:  # type: ignore (not accessed)
             raise CLIException("current organization is not found")
 
         ctx = YataiClientContext(
-            name=default_context_name,
+            name=context,
```
Comment (aarnphm) [inline src/bentoml_cli/yatai.py:49]:
```suggestion
            name=client_context,
```

File: src/bentoml_cli/models.py [hunk 5]
```diff
@@ -243,9 +246,14 @@ def pull(model_tag: str, force: bool):  # type: ignore (not accessed)
...
+        "--context", type=click.STRING, default=None, help="Yatai context name."
+    )
+    def push(model_tag: str, force: bool, threads: int, context: str):  # type: ignore (not accessed)
         """Push Model to a yatai server."""
         model_obj = model_store.get(model_tag)
         if not model_obj:
             raise click.ClickException(f"Model {model_tag} not found in local store")
-        yatai_client.push_model(model_obj, force=force, threads=threads)
+        yatai_client.push_model(
+            model_obj, force=force, threads=threads, context=context
```
Comment (aarnphm) [inline src/bentoml_cli/models.py:258]:
```suggestion
            model_obj, force=force, threads=threads, context=client_context
```

File: src/bentoml_cli/models.py [hunk 6]
```diff
@@ -243,9 +246,14 @@ def pull(model_tag: str, force: bool):  # type: ignore (not accessed)
         default=10,
         help="Number of threads to use for upload",
     )
-    def push(model_tag: str, force: bool, threads: int):  # type: ignore (not accessed)
+    @click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
+    )
+    def push(model_tag: str, force: bool, threads: int, context: str):  # type: ignore (not accessed)
```
Comment (aarnphm) [inline src/bentoml_cli/models.py:252]:
```suggestion
    def push(model_tag: str, force: bool, threads: int, client_context: str):  # type: ignore (not accessed)
```

File: src/bentoml_cli/models.py [hunk 7]
```diff
@@ -243,9 +246,14 @@ def pull(model_tag: str, force: bool):  # type: ignore (not accessed)
         default=10,
         help="Number of threads to use for upload",
     )
-    def push(model_tag: str, force: bool, threads: int):  # type: ignore (not accessed)
+    @click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
```
Comment (aarnphm) [inline src/bentoml_cli/models.py:250]:
```suggestion
        "--client-context", type=click.STRING, default=None, help="Yatai context name."
```

File: src/bentoml_cli/models.py [hunk 8]
```diff
@@ -224,9 +224,12 @@ def import_from(model_path: str) -> None:  # type: ignore (not accessed)
...
         help="Force pull from yatai to local and overwrite even if it already exists in local",
     )
-    def pull(model_tag: str, force: bool):  # type: ignore (not accessed)
+    @click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
+    )
+    def pull(model_tag: str, force: bool, context: str):  # type: ignore (not accessed)
         """Pull Model from a yatai server."""
-        yatai_client.pull_model(model_tag, force=force)
+        yatai_client.pull_model(model_tag, force=force, context=context)
```
Comment (aarnphm) [inline src/bentoml_cli/models.py:232]:
```suggestion
        yatai_client.pull_model(model_tag, force=force, context=client_context)
```

File: src/bentoml_cli/models.py [hunk 9]
```diff
@@ -224,9 +224,12 @@ def import_from(model_path: str) -> None:  # type: ignore (not accessed)
         default=False,
         help="Force pull from yatai to local and overwrite even if it already exists in local",
     )
-    def pull(model_tag: str, force: bool):  # type: ignore (not accessed)
+    @click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
```
Comment (aarnphm) [inline src/bentoml_cli/models.py:228]:
```suggestion
        "--client-context", type=click.STRING, default=None, help="Yatai context name."
```

File: src/bentoml_cli/bentos.py [hunk 10]
```diff
@@ -252,12 +255,17 @@ def pull(bento_tag: str, force: bool) -> None:  # type: ignore (not accessed)
...
+        "--context", type=click.STRING, default=None, help="Yatai context name."
+    )
+    def push(bento_tag: str, force: bool, threads: int, context: str) -> None:  # type: ignore (not accessed)
         """Push Bento to a yatai server."""
         bento_obj = bento_store.get(bento_tag)
         if not bento_obj:
             raise click.ClickException(f"Bento {bento_tag} not found in local store")
-        yatai_client.push_bento(bento_obj, force=force, threads=threads)
+        yatai_client.push_bento(
+            bento_obj, force=force, threads=threads, context=context
```
Comment (aarnphm) [inline src/bentoml_cli/bentos.py:267]:
```suggestion
            bento_obj, force=force, threads=threads, context=client_context
```

File: src/bentoml_cli/bentos.py [hunk 11]
```diff
@@ -252,12 +255,17 @@ def pull(bento_tag: str, force: bool) -> None:  # type: ignore (not accessed)
         default=10,
         help="Number of threads to use for upload",
     )
-    def push(bento_tag: str, force: bool, threads: int) -> None:  # type: ignore (not accessed)
+    @click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
```
Comment (aarnphm) [inline src/bentoml_cli/bentos.py:259]:
```suggestion
        "--client-context", type=click.STRING, default=None, help="Yatai context name."
```

File: src/bentoml_cli/bentos.py [hunk 12]
```diff
@@ -252,12 +255,17 @@ def pull(bento_tag: str, force: bool) -> None:  # type: ignore (not accessed)
...
+    )
+    def push(bento_tag: str, force: bool, threads: int, context: str) -> None:  # type: ignore (not accessed)
```
Comment (aarnphm) [inline src/bentoml_cli/bentos.py:261]:
```suggestion
    def push(bento_tag: str, force: bool, threads: int, client_context: str) -> None:  # type: ignore (not accessed)
```

File: src/bentoml_cli/bentos.py [hunk 13]
```diff
@@ -233,9 +233,12 @@ def import_bento_(bento_path: str) -> None:  # type: ignore (not accessed)
...
         help="Force pull from yatai to local and overwrite even if it already exists in local",
     )
-    def pull(bento_tag: str, force: bool) -> None:  # type: ignore (not accessed)
+    @click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
+    )
+    def pull(bento_tag: str, force: bool, context: str) -> None:  # type: ignore (not accessed)
         """Pull Bento from a yatai server."""
-        yatai_client.pull_bento(bento_tag, force=force)
+        yatai_client.pull_bento(bento_tag, force=force, context=context)
```
Comment (aarnphm) [inline src/bentoml_cli/bentos.py:241]:
```suggestion
        yatai_client.pull_bento(bento_tag, force=force, context=client_context)
```

File: src/bentoml_cli/bentos.py [hunk 14]
```diff
@@ -233,9 +233,12 @@ def import_bento_(bento_path: str) -> None:  # type: ignore (not accessed)
         default=False,
         help="Force pull from yatai to local and overwrite even if it already exists in local",
     )
-    def pull(bento_tag: str, force: bool) -> None:  # type: ignore (not accessed)
+    @click.option(
+        "--context", type=click.STRING, default=None, help="Yatai context name."
```
Comment (aarnphm) [inline src/bentoml_cli/bentos.py:237]:
```suggestion
        "--client-context", type=click.STRING, default=None, help="Yatai context name."
```

File: src/bentoml_cli/bentos.py [hunk 15]
```diff
@@ -233,9 +233,12 @@ def import_bento_(bento_path: str) -> None:  # type: ignore (not accessed)
...
+    )
+    def pull(bento_tag: str, force: bool, context: str) -> None:  # type: ignore (not accessed)
```
Comment (aarnphm) [inline src/bentoml_cli/bentos.py:239]:
```suggestion
    def pull(bento_tag: str, force: bool, client_context: str) -> None:  # type: ignore (not accessed)
```

**Review by aarnphm — COMMENTED** [review]
As discussed.

**Pull request merged**